walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,103 @@
 
 package com.google.sps;
 
+import java.util.Arrays;
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
 
 public final class FindMeetingQuery {
-  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
-  }
+    public static final int DAY_START = TimeRange.START_OF_DAY;
+    public static final int DAY_END = TimeRange.END_OF_DAY;
+    public static final TimeRange WHOLE_DAY = TimeRange.WHOLE_DAY;
+
+    /**
+      * Return a query of TimeRange that attendees can meet for a requested meeting.
+      * If there is not enough room for optional attendees, find TimeRange that just 
+      * work for required attendees. If there is no required attendees, find the query
+      * for just optional attendees.
+      */
+    public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
+        long meetingDuration = request.getDuration();
+        // Required attendees for the request meeting.
+        Collection<String> requiredAttendees = request.getAttendees(); 
+        // All attendees (required and optional) for the request meeting.
+        Collection<String> allAttendees = new ArrayList<>(requiredAttendees);
+        allAttendees.addAll(request.getOptionalAttendees());
+
+        // If the meeting duration is invalid, return an empty Collection.
+        if (meetingDuration > WHOLE_DAY.duration() || meetingDuration == 0) {
+            return Arrays.asList();
+        }
+
+        // Find available TimeRange for both required and optional attendees.
+        Collection<TimeRange> allMeetingTimes = 
+            findMeetingTimes(events, allAttendees, meetingDuration);
+
+        // If there is meeting time for all, return allMeetingTimes so that 
+        // both required and optional can attend the meeting. If there is no 
+        // required attendee, return allMeetingTimes for optional attendees.
+        if (allMeetingTimes.size() > 0 || requiredAttendees.size() == 0) {
+            return allMeetingTimes;
+        }
+        // If not, return available TimeRange for required attendees.
+        return findMeetingTimes(
+            events, requiredAttendees, meetingDuration);
+    }
+
+    /**
+      * Find all times that the meeting could happen, given attendees and duration. 
+      * Pointer rangeStart is pointing at the start of the current available 
+      * TimeRange to keep track of where the meeting could possibly start. The time 
+      * between the rangeStart and start time of the current event is the current 
+      * avaliable TimeRange. If the current available TimeRange has longer duration 
+      * than the requested meeting duration, add it to available meetingTimes.
```

Comment:
This comment still does not clearly describes the algorithm. For example, it doesn't mention that the events will be sorted and why doing it is helpful. It's also not clear how the `rangeStart` will be updated.

Reply:
The comment is updated! :)